File: core/src/ru/alexkulikov/firstfame/levels/LevelBuilder.java
package ru.alexkulikov.firstfame.levels;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.math.Circle;
import com.badlogic.gdx.math.Intersector;
import com.badlogic.gdx.math.Polygon;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.World;

import ru.alexkulikov.firstfame.PolygonUtils;
import ru.alexkulikov.firstfame.objects.Box;
import ru.alexkulikov.firstfame.objects.BoxData;
import ru.alexkulikov.firstfame.objects.Material;
import ru.alexkulikov.firstfame.objects.ObjectType;
import ru.alexkulikov.firstfame.objects.Platform;

import com.badlogic.gdx.scenes.scene2d.Group;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.XmlReader;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class LevelBuilder {

    private World world;
    private XmlReader reader;
    private List<Polygon> contactPlatforms;
    private Group levelGroup;
    private PolygonUtils polygonUtils;

    public LevelBuilder(World world) {
        this.world = world;
        reader= new XmlReader();
        polygonUtils = new PolygonUtils();
    }

    public void buildGroups(String levelName, LevelBuiltCallback callback) {
        levelGroup = new Group();
        contactPlatforms = new ArrayList<Polygon>();

//        try {
            XmlReader.Element root = reader.parse(Gdx.files.internal("levels/" + levelName));
            XmlReader.Element player = root.getChildByName("player");

            float playerX = Float.parseFloat(player.getAttribute("x"));
            float playerY = Float.parseFloat(player.getAttribute("y"));

            XmlReader.Element boxes = root.getChildByName("boxes");
            int boxCount = boxes.getChildCount();
            for (int i = 0; i < boxCount; i++) {
                XmlReader.Element box = boxes.getChild(i);
                String type = box.getName();

                float x = Float.parseFloat(box.getAttribute("x"));
                float y = Float.parseFloat(box.getAttribute("y"));
                float w = Float.parseFloat(box.getAttribute("w"));
                float h = Float.parseFloat(box.getAttribute("h"));
                Material material = Material.valueOf(box.getAttribute("material"));

                if (type.equals("box")) {
                    levelGroup.addActor(new Box(world, material, x, y, w, h));
                } else {
                    Platform platform = new Platform(world, material, x, y, w, h);
                    contactPlatforms.add(platform.getContactPolygon());
                    levelGroup.addActor(platform);
                }
            }

            callback.onBuilt(playerX, playerY);

//        } catch (IOException e) {
//            Gdx.app.log("Level builder", "Cannot parse xml", e);
//        }
    }

    public void clearBoxesBodies() {
        Array<Body> bodies = new Array<Body>();
        world.getBodies(bodies);
        for (Body b : bodies) {
            BoxData data = (BoxData) b.getUserData();
            if (data != null && (data.getType() == ObjectType.box || data.getType() == ObjectType.player)) {
                world.destroyBody(b);
            }
        }
    }

    public void clearLevel() {
        levelGroup.remove();
        contactPlatforms = new ArrayList<Polygon>();
        clearBoxesBodies();
    }

    public boolean onPlatform(Polygon playerPolygon) {
        for (Polygon polygon : contactPlatforms) {
            if (Intersector.overlapConvexPolygons(polygon, playerPolygon)) {
                return true;
            }
        }

        return false;
    }

    public boolean onPlatform(Circle playerPolygon) {
        for (Polygon polygon : contactPlatforms) {
            if (polygonUtils.overlaps(polygon, playerPolygon)) {
                return true;
            }
        }

        return false;
    }

    public Group getLevelGroup() {
        return levelGroup;
    }
}

File: core/src/ru/alexkulikov/firstfame/levels/LevelBuilder.kt
package ru.alexkulikov.firstfame.levels

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.math.Polygon
import com.badlogic.gdx.physics.box2d.Body
import com.badlogic.gdx.physics.box2d.BodyDef
import com.badlogic.gdx.physics.box2d.Joint
import com.badlogic.gdx.physics.box2d.World
import com.badlogic.gdx.scenes.scene2d.Group
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.XmlReader
import ru.alexkulikov.firstfame.Path
import ru.alexkulikov.firstfame.objects.*
import ru.alexkulikov.firstfame.objects.player.Player
import ru.alexkulikov.firstfame.objects.player.QuadPlayer

class LevelBuilder(private val world: World, private val manager: AssetManager) {
    private val reader = XmlReader()
    private var contactPlatforms = mutableListOf<Polygon>()
    val zooms = mutableListOf<Zoom>()

    private lateinit var levelGroup: Group
    lateinit var playerActor: Player
    private lateinit var exitActor: Exit

    private val woodTexture: Texture = manager.get(Path.woodMaterial)
    private val iceTexture: Texture = manager.get(Path.iceMaterial)

    init {
        woodTexture.setWrap(Texture.TextureWrap.Repeat, Texture.TextureWrap.Repeat)
        iceTexture.setWrap(Texture.TextureWrap.Repeat, Texture.TextureWrap.Repeat)
    }

    fun build(levelName: String, stage: Stage) {
        contactPlatforms.clear()
        zooms.clear()
        levelGroup = Group()

        val root = reader.parse(Gdx.files.internal("levels/" + levelName))

        buildBoxes(root)
        stage.addActor(levelGroup)

        parseZoom(root)

        val exit = root.getChildByName("exit")
        val exitPos = parseCoordinates(exit)
        exitActor = Exit(manager, exitPos.first, exitPos.second)
        stage.addActor(exitActor)

        val player = root.getChildByName("player")
        val playerPos = parseCoordinates(player)
        playerActor = QuadPlayer(world, playerPos.first, playerPos.second, 0.4f, 0.4f)
        stage.addActor(playerActor)
    }

    private fun parseCoordinates(element: XmlReader.Element): Pair<Float, Float> {
        return Pair(element.getFloatAttribute("x"), element.getFloatAttribute("y"))
    }

    private fun parseZoom(root: XmlReader.Element) {
        val zooms = root.getChildByName("zooms") ?: return

        val zoomCount = zooms.childCount

        (0 until zoomCount).forEach {
            val zoom = zooms.getChild(it)

            val x = zoom.getFloatAttribute("x")
            val y = zoom.getFloatAttribute("y")
            val w = zoom.getFloatAttribute("w")
            val h = zoom.getFloatAttribute("h")
            val scale = zoom.getFloatAttribute("scale")

            this.zooms.add(Zoom(x, y, h, w, scale))
        }
    }

    private fun buildBoxes(root: XmlReader.Element) {
        val boxes = root.getChildByName("boxes")
        val boxCount = boxes.childCount

        (0 until boxCount).forEach {
            val box = boxes.getChild(it)
            val type = box.name

            val x = box.getFloatAttribute("x")
            val y = box.getFloatAttribute("y")
            val w = box.getFloatAttribute("w")
            val h = box.getFloatAttribute("h")
            val material = Material.valueOf(box.getAttribute("material"))

            val materialTexture = when (material) {
                Material.wood -> woodTexture
                Material.ice -> iceTexture
                else -> {
                    woodTexture
                }
            }

            when (type) {
                "box" -> {
                    val platform = Platform(materialTexture, world, material, x, y, w, h)
                    contactPlatforms.add(platform.contactPolygon)
                    levelGroup.addActor(platform)
                }
                "ground" -> {
                    val platform = Platform(materialTexture, world, material, x, y, w, h, BodyDef.BodyType.StaticBody)
                    contactPlatforms.add(platform.contactPolygon)
                    levelGroup.addActor(platform)
                }
                "seesaw" -> {
                    val platform = Platform(materialTexture, world, ObjectType.SEESAW, material, x, y, w, h)
                    contactPlatforms.add(platform.contactPolygon)
                    levelGroup.addActor(platform)
                }
                "circle" -> levelGroup.addActor(CircleBox(world, material, x, y, h))
            }
        }
    }

    private fun clearWorld() {
        val bodies = Array<Body>()
        val joints = Array<Joint>()
        world.getBodies(bodies)
        world.getJoints(joints)
        joints.forEach { world.destroyJoint(it) }
        bodies.forEach {
            val data = it.userData as BoxData?
            if (data == null || data.type != ObjectType.GROUND) {
                world.destroyBody(it)
            }
        }
    }

    fun clearLevel() {
        levelGroup.remove()
        exitActor.remove()
        playerActor.remove()
        clearWorld()
    }

    fun onPlatform(player: Player): Boolean =
            contactPlatforms
                    .filter { it.x > player.x - 4 && it.x < player.x + 4}
                    .any { player.overlaps(it) }
}